File: server/src/engine/entitymanager.ts
//SUGGESTION: Transform it to system?
//TODO: make it use data from DB
class EntityManager {
    private _lowestUnassignedEntityID: number;
    private _entityStore: number[];
    private _entityHumanReadableNames: { [entity: number]: string };
    private _componentStore: { [name: string]: { [entity: number]: Component } };

    constructor() {
        this._lowestUnassignedEntityID = 1;
        this._entityStore = [];
        this._componentStore = {};
        this._entityHumanReadableNames = [];
    }

    private _generateNewEntityID(): number {
        //TODO: Make it synchronous
        {
            if (this._lowestUnassignedEntityID < Number.MAX_VALUE) {
                return this._lowestUnassignedEntityID++;
            }
            else {
                for (var i = 1; i < Number.MAX_VALUE; i++) {
                    if (this._entityStore.indexOf(i) != -1)
                        return i;
                }
                throw new Error("ERROR: no available Entity IDs; too many entities!");
            }
        }
    }

    createEntity(name?: string): number {
        var entity = this._generateNewEntityID();
        if (entity < 1) {
            throw new Error("WTF? Entity < 1");
        }
        else {
            this._entityStore.push(entity);
            if (name)
                this._entityHumanReadableNames[entity] = name;
            return entity;
        }
    }

    addComponent(entity: number, component: Component): void {
        var store = this._componentStore[component.componentName];
        if (store == undefined) {
            store = {};
            this._componentStore[component.componentName] = store;
        }
        store[entity] = component;
    }

    addComponents(entity: number, components: Component[]): void {
        for (var i = 0; i < components.length; i++) {
            this.addComponent(entity, components[i]);
        }
    }

    hasComponent<T extends Component>(entity: number) {
        var store = this._componentStore[typeof T];
        if (store == undefined)
            return false;
        else
            return store[entity] != undefined;
    }

    getComponent<T extends Component>(entity: number, componentType: { prototype: T }): T {
        var store = this._componentStore[componentType.prototype.getType()];
        if (store == undefined)
            throw new Error("GET FAIL: there are no entities with a Component of class: " + componentType.prototype.getType());
        var result: T = <T> store[entity];
        if (result == undefined)
            throw new Error("GET FAIL: " + this.nameFor(entity) + " does not possess Component of class\n missing: " + componentType.prototype.getType());
        return result;
    }

    getAllComponentsOfType<T extends Component>(componentType: { prototype: T }): T[] {
        var store = this._componentStore[componentType.prototype.getType()];
        var result = [];
        if (store != undefined)
            for (var entity in store)
                result.push(store[entity]);
        return result;
    }

    getAllEntitiesPossessingComponent<T extends Component>(componentType: { prototype: T }): number[] {
        var store = this._componentStore[componentType.prototype.getType()];
        var result = [];
        if (store != undefined)
            for (var entity in store)
                result.push(entity);
        return result;
    }

    getAllComponentsOnEntity<T extends Component>(entity: number): T[] {
        var components: T[] = [];
        for (var componentType in this._componentStore) {
            var component = this._componentStore[componentType][entity];
            if (component == undefined)
                continue;
            components.push(<T>component);
        }
        return components;
    }

    removeComponent<T extends Component>(entity: number, componentType: { prototype: T }): void {
        var store = this._componentStore[componentType.prototype.getType()];
        if (store == undefined)
            throw new Error("REMOVE FAIL: there are no entities with a Component of class: " + componentType.prototype.getType());
        store[entity] = undefined;
    }

    removeComponents<T extends Component>(entity: number, componentTypes: { prototype: T }[]): void {
        for (var i = 0; i < componentTypes.length; i++) {
            this.removeComponent(entity, componentTypes[i]);
        }
    }

    destroyEntity(entity: number): void {
        //TODO: Make it synchronous
        var index = this._entityStore.indexOf(entity);
        if (index == -1) {
            return;
        }
        this._entityStore.splice(index, 1);
        if (entity == this._lowestUnassignedEntityID)
            this._lowestUnassignedEntityID--;
        for (var componentType in this._componentStore) {
            this._componentStore[componentType][entity] = undefined;
        }
    }

    destroyEntities(entities: number[]) {
        for (var i = 0; i < entities.length; i++) {
            this.destroyEntity(entities[i]);
        }
    }

    setEntityName(entity: number, name: string): void {
        this._entityHumanReadableNames[entity] = name;
    }

    nameFor(entity: number): string {
        return this._entityHumanReadableNames[entity];
    }
}

export = EntityManager;

File: sharpServer/Engine/entitymanager.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SharpServer.Engine
{
    using ComponentStore = Dictionary<uint, Component>;

    // SUGGESTION: Transform it to service?
    // TODO: make it use data from DB
    public static class EntityManager
    {
        private static readonly Dictionary<string, ComponentStore> ComponentDb =
            new Dictionary<string, ComponentStore>();

        private static readonly Dictionary<uint, string> EntityHumanReadableNames = new Dictionary<uint, string>();
        private static readonly List<uint> EntityStore = new List<uint>();
        private static uint lowestUnassignedEntityId = 1;

        public static void AddComponent<T>(uint entity, T component) where T : Component
        {
            ComponentStore store;
            if (!ComponentDb.TryGetValue(GetTypeName<T>(), out store))
            {
                store = new ComponentStore();
                ComponentDb[GetTypeName<T>()] = store;
            }
            store[entity] = component;
        }

        public static void AddComponents(uint entity, params Component[] components)
        {
            foreach (var component in components)
            {
                AddComponent(entity, component);
            }
        }

        public static uint CreateEntity(string name = "No description")
        {
            var entity = GenerateNewEntityId();
            EntityStore.Add(entity);
            EntityHumanReadableNames[entity] = name;
            return entity;
        }

        public static void DestroyEntities(params uint[] entities)
        {
            foreach (var entity in entities)
                DestroyEntity(entity);
        }

        // TODO: Make it synchronous
        public static void DestroyEntity(uint entity)
        {
            if (!EntityStore.Contains(entity))
                return;
            EntityStore.Remove(entity);
            foreach (var componentType in ComponentDb)
            {
                ComponentDb[componentType.Key].Remove(entity);
            }
        }

        public static T[] GetAllComponentsOfType<T>() where T : Component
        {
            return ComponentDb[GetTypeName<T>()]
                .Values
                .Cast<T>()
                .ToArray();
        }

        // WARNING: Use only for debug reasons (slow as fuck)
        public static T[] GetAllComponentsOnEntity<T>(uint entity) where T : Component
        {
            return ComponentDb
                .Where(componentType => componentType.Value.ContainsKey(entity))
                .Select(componentType => componentType.Value[entity])
                .Cast<T>()
                .ToArray();
        }

        public static uint[] GetAllEntitiesPossessingComponent<T>() where T : Component
        {
            return ComponentDb[GetTypeName<T>()]
                .Keys
                .ToArray();
        }

        public static T GetComponent<T>(uint entity) where T : Component
        {
            ComponentStore store;
            if (!ComponentDb.TryGetValue(GetTypeName<T>(), out store))
                throw new KeyNotFoundException("GET FAIL: there are no entities with a Component of class: " +
                                               GetTypeName<T>());
            if (!store.ContainsKey(entity))
                throw new KeyNotFoundException("GET FAIL: " + NameFor(entity) + "(ID: " + entity +
                                               ") does not possess Component: " + GetTypeName<T>());
            return store[entity] as T;
        }

        private static string GetTypeName<T>() where T : Component
        {
            return typeof (T).Name;
        }

        public static bool TryGetComponent<T>(uint entity, out T component) where T : Component
        {
            ComponentStore store;
            component = null;
            if (!ComponentDb.TryGetValue(GetTypeName<T>(), out store))
                return false;
            if (!store.ContainsKey(entity))
                return false;
            component = store[entity] as T;
            return true;
        }

        public static bool HasComponent<T>(uint entity) where T : Component
        {
            ComponentStore store;
            return ComponentDb.TryGetValue(GetTypeName<T>(), out store) && store.ContainsKey(entity);
        }

        public static string NameFor(uint entity)
        {
            return EntityHumanReadableNames.ContainsKey(entity) ? EntityHumanReadableNames[entity] : "Not defined";
        }

        public static void RemoveComponent<T>(uint entity) where T : Component
        {
            ComponentStore store;
            if (ComponentDb.TryGetValue(GetTypeName<T>(), out store))
                store.Remove(entity);
        }

        public static void SetEntityName(uint entity, string name)
        {
            EntityHumanReadableNames[entity] = name;
        }

        // SUGGESTION: Think about better implementation
        // TODO: Make it synchronous
        private static uint GenerateNewEntityId()
        {
            if (lowestUnassignedEntityId < int.MaxValue)
            {
                return lowestUnassignedEntityId++;
            }
            for (uint i = 1; i < int.MaxValue; i++)
            {
                if (!EntityStore.Contains(i))
                    return i;
            }
            throw new ArgumentOutOfRangeException("ERROR: no available Entity IDs; too many entities!");
        }
    }
}